src/device-api/common.js

```diff
@@ -0,0 +1,88 @@
+import * as Bluebird from 'bluebird';
+import * as _ from 'lodash';
+import { appNotFoundMessage } from '../lib/messages';
+
+export function doRestart(applications, appId, force) {
+	const { _lockingIfNecessary, deviceState } = applications;
+
+	return _lockingIfNecessary(appId, { force }, () =>
+		deviceState.getCurrentForComparison().then(function(currentState) {
+			const app = currentState.local.apps[appId];
+			const imageIds = _.map(app.services, 'imageId');
+			applications.clearTargetVolatileForServices(imageIds);
+			const stoppedApp = _.cloneDeep(app);
+			stoppedApp.services = [];
+			currentState.local.apps[appId] = stoppedApp;
+			return deviceState
+				.pausingApply(() =>
+					deviceState
+						.applyIntermediateTarget(currentState, { skipLock: true })
+						.then(function() {
+							currentState.local.apps[appId] = app;
+							return deviceState.applyIntermediateTarget(currentState, {
+								skipLock: true,
+							});
+						}),
+				)
+				.finally(() => deviceState.triggerApplyTarget());
+		}),
+	);
+}
+
+export function doPurge(applications, appId, force) {
+	const { logger, _lockingIfNecessary, deviceState, volumes } = applications;
+
+	logger.logSystemMessage(
+		`Purging data for app ${appId}`,
+		{ appId },
+		'Purge data',
+	);
+	return _lockingIfNecessary(appId, { force }, () =>
+		deviceState.getCurrentForComparison().then(function(currentState) {
+			const app = currentState.local.apps[appId];
+			if (app == null) {
+				throw new Error(appNotFoundMessage);
+			}
+			const purgedApp = _.cloneDeep(app);
+			purgedApp.services = [];
+			purgedApp.volumes = {};
+			currentState.local.apps[appId] = purgedApp;
+			return deviceState
+				.pausingApply(() =>
+					deviceState
+						.applyIntermediateTarget(currentState, { skipLock: true })
+						.then(() =>
+							// Now that we're not running anything, explicitly
+							// remove the volumes, we must do this here, as the
+							// application-manager will not remove any volumes
+							// which are part of an active application
+							Bluebird.each(volumes.getAllByAppId(appId), vol => vol.remove()),
```

Comment:
It looks like you're missing a return here?

Reply:
And in fact I would prefer to do this all in parallel.

Reply:
It's an implicit return function, just the comment makes that hard to spot - I'll make it an explicit return function instead. 
For parallel vs serial I'd rather leave it as-is to keep the same behavior during the conversion and make that change in a separate PR since it could bring up edge cases bugs

Reply:
I changed this to an explicit return

Reply:
I'm happy enough leaving it as-is for now.